crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -100,36 +104,49 @@ impl Trade {
     /// fees have been applied and calculated over the price limits.
     ///
     /// Denominated in SURPLUS token
-    fn surplus(&self, price_limits: PriceLimits) -> Option<eth::Asset> {
+    fn surplus_over_reference_price(&self, price_limits: PriceLimits) -> Result<eth::Asset, Error> {
         match self.side {
             Side::Buy => {
                 // scale limit sell to support partially fillable orders
                 let limit_sell = price_limits
                     .sell
                     .0
-                    .checked_mul(self.executed.into())?
-                    .checked_div(price_limits.buy.0)?;
-                // difference between limit sell and executed amount converted to sell token
-                limit_sell.checked_sub(
-                    self.executed
-                        .0
-                        .checked_mul(self.custom_price.buy)?
-                        .checked_div(self.custom_price.sell)?,
-                )
+                    .checked_mul(self.executed.into())
+                    .ok_or(Math::Overflow)?
+                    .checked_div(price_limits.buy.0)
+                    .ok_or(Math::DivisionByZero)?;
+                let sold = self
+                    .executed
+                    .0
+                    .checked_mul(self.custom_price.buy)
+                    .ok_or(Math::Overflow)?
+                    .checked_div(self.custom_price.sell)
+                    .ok_or(Math::DivisionByZero)?;
+                limit_sell.checked_sub(sold).ok_or(Error::NegativeSurplus(
```

Comment:
Wondering if this should be `Math::Negative` (or `Math::Underflow`) instead?

Reply:
Wasn't sure about this one, since it could lead to accidentally adding another case in the future, where `Math::Negative` is returned, while not being related to negative surplus... but implemented, let's keep the current code clean.